unittests/unittest_mtrie.cpp

```diff
@@ -0,0 +1,438 @@
+    const zmq::generic_mtrie_t<int>::prefix_t test_name =
+      reinterpret_cast<zmq::generic_mtrie_t<int>::prefix_t> ("foo");
+
+    int count = 0;
+    mtrie.match (test_name, getlen (test_name), mtrie_count, &count);
+    TEST_ASSERT_EQUAL_INT (0, count);
+}
+
+void test_check_empty_match_empty_data ()
+{
+    zmq::generic_mtrie_t<int> mtrie;
+
+    int count = 0;
+    mtrie.match (NULL, 0, mtrie_count, &count);
+    TEST_ASSERT_EQUAL_INT (0, count);
+}
+
+void test_add_single_entry_match_exact ()
+{
+    int pipe;
+
+    zmq::generic_mtrie_t<int> mtrie;
+    const zmq::generic_mtrie_t<int>::prefix_t test_name =
+      reinterpret_cast<zmq::generic_mtrie_t<int>::prefix_t> ("foo");
+
+    bool res = mtrie.add (test_name, getlen (test_name), &pipe);
+    TEST_ASSERT_TRUE (res);
+
+    int count = 0;
+    mtrie.match (test_name, getlen (test_name), mtrie_count, &count);
+    TEST_ASSERT_EQUAL_INT (1, count);
+}
+
+void test_add_two_entries_with_same_name_match_exact ()
+{
+    int pipe_1, pipe_2;
+
+    zmq::generic_mtrie_t<int> mtrie;
+    const zmq::generic_mtrie_t<int>::prefix_t test_name =
+      reinterpret_cast<zmq::generic_mtrie_t<int>::prefix_t> ("foo");
+
+    bool res = mtrie.add (test_name, getlen (test_name), &pipe_1);
+    TEST_ASSERT_TRUE (res);
+
+    res = mtrie.add (test_name, getlen (test_name), &pipe_2);
+    TEST_ASSERT_FALSE (res);
+
+    int count = 0;
+    mtrie.match (test_name, getlen (test_name), mtrie_count, &count);
+    TEST_ASSERT_EQUAL_INT (2, count);
+}
+
+void test_add_two_entries_match_prefix_and_exact ()
+{
+    int pipe_1, pipe_2;
+
+    zmq::generic_mtrie_t<int> mtrie;
+    const zmq::generic_mtrie_t<int>::prefix_t test_name_prefix =
+      reinterpret_cast<zmq::generic_mtrie_t<int>::prefix_t> ("foo");
+    const zmq::generic_mtrie_t<int>::prefix_t test_name_full =
+      reinterpret_cast<zmq::generic_mtrie_t<int>::prefix_t> ("foobar");
+
+    bool res = mtrie.add (test_name_prefix, getlen (test_name_prefix), &pipe_1);
+    TEST_ASSERT_TRUE (res);
+
+    res = mtrie.add (test_name_full, getlen (test_name_full), &pipe_2);
+    TEST_ASSERT_TRUE (res);
+
+    int count = 0;
+    mtrie.match (test_name_full, getlen (test_name_full), mtrie_count, &count);
+    TEST_ASSERT_EQUAL_INT (2, count);
+}
+
+void test_add_rm_single_entry_match_exact ()
+{
+    int pipe;
+    zmq::generic_mtrie_t<int> mtrie;
+    const zmq::generic_mtrie_t<int>::prefix_t test_name =
+      reinterpret_cast<zmq::generic_mtrie_t<int>::prefix_t> ("foo");
+
+    mtrie.add (test_name, getlen (test_name), &pipe);
+    bool res = mtrie.rm (test_name, getlen (test_name), &pipe);
+    TEST_ASSERT_TRUE (res);
+
+    int count = 0;
+    mtrie.match (test_name, getlen (test_name), mtrie_count, &count);
+    TEST_ASSERT_EQUAL_INT (0, count);
+}
+
+void test_rm_nonexistent_0_size_empty ()
+{
+    int pipe;
+    zmq::generic_mtrie_t<int> mtrie;
+
+    // TODO why does this return true, but test_rm_nonexistent_empty returns false?
```

Comment:
The single user of the return value of the mtrie is the verbose(r) XPUB option - unfortunately the documentation of that option is very confusing, and also there is no test for it at all.

I'm trying to understand exactly how the unsub verboser mode works and write tests for it, so that we can establish the boundaries (given by the existing public API behaviour) within which we can move and change things.

Reply:
Ok I think I've covered the most common use cases - I'll push the test to this branch

Reply:
Ok, that sounds great. I will look at the test cases, and see what can be done to mtrie based on them.